src/Simulator/IncrementalRedstoneSimulator/PressurePlateHandler.h

```diff
@@ -72,16 +78,134 @@ class cPressurePlateHandler : public cRedstoneHandler
 
-		auto Power = GetPowerLevel(a_World, a_Position, a_BlockType, a_Meta);
-		auto PreviousPower = static_cast<cIncrementalRedstoneSimulator *>(a_World.GetRedstoneSimulator())->GetChunkData()->ExchangeUpdateOncePowerData(a_Position, PoweringData(a_BlockType, Power));
+		auto ChunkData = static_cast<cIncrementalRedstoneSimulator *>(a_World.GetRedstoneSimulator())->GetChunkData();
+
+		const auto PreviousPower = ChunkData->GetCachedPowerData(a_Position);
+		auto Power = GetPowerLevel(a_World, a_Position, a_BlockType, a_Meta);  // Get the current power of the platey
 
-		if (Power != PreviousPower.PowerLevel)
+		const auto PlateUpdates = GetAdjustedRelatives(a_Position, StaticAppend(GetRelativeLaterals(), cVector3iArray{ OffsetYM() }));
+		auto DelayInfo = ChunkData->GetMechanismDelayInfo(a_Position);
+
+		// Resting state?
+		if (DelayInfo == nullptr)
 		{
-			a_World.SetBlockMeta(a_Position, (Power == 0) ? E_META_PRESSURE_PLATE_RAISED : E_META_PRESSURE_PLATE_DEPRESSED);
-			return GetAdjustedRelatives(a_Position, StaticAppend(GetRelativeLaterals(), cVector3iArray{ OffsetYM() }));
+			if (Power == 0)
+			{
+				// Nothing happened, back to rest
+				return {};
+			}
+
+			// From rest, a player stepped on us
+			// Schedule a minimum 0.5 second delay before even thinking about releasing
+			ChunkData->m_MechanismDelays[a_Position] = std::make_pair(5, true);
+
+			// manage on-sound
+			AString soundToPlay = "";
+			switch (a_BlockType)
+			{
+				case E_BLOCK_STONE_PRESSURE_PLATE:
+					soundToPlay = "block.wood_pressureplate.click_on";
+					break;
+				case E_BLOCK_WOODEN_PRESSURE_PLATE:
+					soundToPlay = "block.wood_pressureplate.click_on";
+					break;
+				case E_BLOCK_HEAVY_WEIGHTED_PRESSURE_PLATE:
+				case E_BLOCK_LIGHT_WEIGHTED_PRESSURE_PLATE:
+					soundToPlay = "block.metal_pressureplate.click_on";
+					break;
+				default:
+				{
+					ASSERT(!"Unhandled/unimplemented block in pressure plate handler!");
+					return {};
+				}
+			}
+			a_World.BroadcastSoundEffect(soundToPlay, a_Position, 0.5f, 0.6f);
+
+			// Update power
+			ChunkData->SetCachedPowerData(a_Position, PoweringData(a_BlockType, Power));
+
+			// Immediately depress plate
+			a_World.SetBlockMeta(a_Position, E_META_PRESSURE_PLATE_DEPRESSED);
+			return PlateUpdates;
 		}
 
-		return {};
+		// Not a resting state
+
+		int DelayTicks;
+		bool HasExitedMinimumOnDelayPhase;
+		std::tie(DelayTicks, HasExitedMinimumOnDelayPhase) = *DelayInfo;
+
+		// Are we waiting for the initial delay or subsequent release delay?
+		if (DelayTicks > 0)
+		{
+			// Yes. Are we waiting to release, and found that the player stepped on it again?
+			if (!HasExitedMinimumOnDelayPhase && (Power > 0))
+			{
+				// Reset delay
+				*DelayInfo = std::make_pair(0, true);
+			}
+
+			// Did the power level change and is still above zero?
+			if ((Power != PreviousPower.PowerLevel) && (Power > 0))
```

Comment:
Instead of putting `&& (Power > 0)` on every condition, how about returning early?
```cpp
if (Power == 0)
{
    return {};
}
```